diff --git a/logrotate.c b/logrotate.c
index be335b0..44a9c89 100644
--- a/logrotate.c
+++ b/logrotate.c
@@ -61,6 +61,11 @@ struct logNames {
     char *baseName;
 };
 
+struct compData {
+	int prefix_len;
+	const char *dformat;
+};
+
 struct logStates {
 	LIST_HEAD(stateSet, logState) head;
 } **states;
@@ -82,6 +87,34 @@ static int globerr(const char *pathname, int theerr)
     return 1;
 }
 
+static int compGlobResult(const void *result1, const void *result2, void *data)  {
+	struct tm time;
+	time_t t1, t2;
+	struct compData *d = (struct compData *) data;
+	const char *r1 = *(const char **)(result1);
+	const char *r2 = *(const char **)(result2);
+
+	memset(&time, 0, sizeof(struct tm));
+	strptime(r1 + d->prefix_len, d->dformat, &time);
+	t1 = mktime(&time);
+
+	memset(&time, 0, sizeof(struct tm));
+	strptime(r2 + d->prefix_len, d->dformat, &time);
+	t2 = mktime(&time);
+
+	if (t1 < t2) return -1;
+	if (t1 > t2) return  1;
+	return 0;
+}
+
+static void sortGlobResult(glob_t *result, int prefix_len, const char *dformat) {
+	struct compData d;
+	if (!dformat || *dformat == '\0') return;
+	d.prefix_len = prefix_len;
+	d.dformat = dformat;
+	qsort_r(result->gl_pathv, result->gl_pathc, sizeof(char *), compGlobResult, &d);
+}
+
 static void unescape(char *arg)
 {
 	char *p = arg;
@@ -807,7 +807,7 @@ int prerotateSingleLog(struct logInfo *log, int logNum, struct logState *state,
 #define DATEEXT_LEN 64
 #define PATTERN_LEN (DATEEXT_LEN * 2)
 	char dext_str[DATEEXT_LEN];
-	char dformat[DATEEXT_LEN];
+	char dformat[DATEEXT_LEN] = "";
 	char dext_pattern[PATTERN_LEN];
 	char *dext;
 
@@ -950,6 +982,7 @@ int prerotateSingleLog(struct logInfo *log, int logNum, struct logState *state,
 		}
 	    rc = glob(glob_pattern, 0, globerr, &globResult);
 	    if (!rc && globResult.gl_pathc > 0) {
+		sortGlobResult(&globResult, strlen(rotNames->dirName) + 1 + strlen(rotNames->baseName), dformat);
 		for (i = 0; i < globResult.gl_pathc && !hasErrors; i++) {
 		    struct stat sbprev;
 
@@ -1006,6 +1039,7 @@ int prerotateSingleLog(struct logInfo *log, int logNum, struct logState *state,
 	    /* remove the first (n - rotateCount) matches
 	     * no real rotation needed, since the files have
 	     * the date in their name */
+		sortGlobResult(&globResult, strlen(rotNames->dirName) + 1 + strlen(rotNames->baseName), dformat);
 	    for (i = 0; i < globResult.gl_pathc; i++) {
 		if (!stat((globResult.gl_pathv)[i], &fst_buf)) {
 		    if ((i <= ((int) globResult.gl_pathc - rotateCount))
