Index: /trunk/logrotate.c
===================================================================
--- trunk/logrotate.c	(revision 368)
+++ trunk/logrotate.c	(revision 369)
@@ -1123,5 +1123,6 @@
 	/* firstRotated is most recently created/compressed rotated log */
 	sprintf(rotNames->firstRotated, "%s/%s%s%s%s",
-		rotNames->dirName, rotNames->baseName, dext_str, fileext, compext);
+		rotNames->dirName, rotNames->baseName, dext_str, fileext,
+		(log->flags & LOG_FLAG_DELAYCOMPRESS) ? "" : compext);
 	globfree(&globResult);
 	free(glob_pattern);
